File: app/src/main/java/com/swipecare/payments/recharges/Mobile_Recharge.java
package com.swipecare.payments.recharges;

import android.app.Activity;
import android.app.AlertDialog;
import android.app.ProgressDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.SharedPreferences;

import android.database.Cursor;
import android.net.Uri;
import android.os.AsyncTask;
import android.os.Bundle;
import android.provider.ContactsContract;
import androidx.annotation.Nullable;
import androidx.fragment.app.FragmentManager;
import androidx.appcompat.app.AppCompatActivity;
import android.util.Log;
import android.view.MenuItem;
import android.view.View;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.RelativeLayout;
import android.widget.TextView;
import android.widget.Toast;

import com.swipecare.payments.DetectConnection;
import com.swipecare.payments.Operators.OperatorsFragment;
import com.swipecare.payments.R;
import com.swipecare.payments.SharePrfeManager;
import com.swipecare.payments.TransactionReciept;


import org.json.JSONException;
import org.json.JSONObject;

import java.io.BufferedInputStream;
import java.io.BufferedReader;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;

public class Mobile_Recharge extends AppCompatActivity  {

    LinearLayout ll_operator;
    TextView textview_operator;


    EditText editext_number;
    public static  EditText edittext_amount;
    Button button_paynow;

    ProgressDialog dialog;

    String myJSON;

    String username,password,balance,provider="";


    public String amount="",number;
    RelativeLayout rl_message;
    TextView textview_message;


    ImageView imageview_contact;

    String operator_name,provider_image;

    LinearLayout ll_dth,ll_electricity;

    String icon="";

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

    // Inflate the layout for this fragment
        setContentView(R.layout.mobile_recharges);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        ll_dth=findViewById(R.id.ll_dth);
        ll_dth.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                startActivity(new Intent(Mobile_Recharge.this,DTH.class));
                Mobile_Recharge.this.overridePendingTransition(android.R.anim.slide_in_left, android.R.anim.slide_out_right);
                finish();
            }
        });

        ll_electricity=findViewById(R.id.ll_electricity);
        ll_electricity.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                startActivity(new Intent(Mobile_Recharge.this,Electricity.class));
                Mobile_Recharge.this.overridePendingTransition(android.R.anim.slide_in_left, android.R.anim.slide_out_right);
                finish();
            }
        });

        SharedPreferences sharedPreferences = getSharedPreferences("user", Context.MODE_PRIVATE);
        username=sharedPreferences.getString("username","");
        password=sharedPreferences.getString("password","");
        balance=sharedPreferences.getString("balance","");


        imageview_contact=(ImageView)findViewById(R.id.imageview_contact);
        imageview_contact.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent intent = new Intent(Intent.ACTION_PICK, ContactsContract.CommonDataKinds.Phone.CONTENT_URI);
                startActivityForResult(intent, 10);
            }
        });


        button_paynow = findViewById(R.id.button_paynow);
        editext_number = findViewById(R.id.edittext_enter_number);
        edittext_amount = (EditText) findViewById(R.id.edittext_amount);

        rl_message= findViewById(R.id.rl_message);
        textview_message= findViewById(R.id.textview_message);


        ll_operator= findViewById(R.id.ll_operator);
        textview_operator= findViewById(R.id.textview_operator);
        ll_operator.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                OperatorsFragment bottomSheetDialogFragment = new OperatorsFragment();
                Bundle bundle=new Bundle();
                bundle.putString("type","1");
                bundle.putString("activity","mobile");
                bottomSheetDialogFragment.setArguments(bundle);

                FragmentManager fragmentManager=getSupportFragmentManager();
                getSupportFragmentManager().beginTransaction().add(R.id.content,bottomSheetDialogFragment).addToBackStack(null).commit();

            }
        });


        button_paynow.setOnClickListener(new View.OnClickListener() {

            @Override
            public void onClick(View v) {

                if (DetectConnection.checkInternetConnection(Mobile_Recharge.this)) {
                    if (editext_number.getText().toString().equals("")) {
                        textview_message.setText("Please enter mobileno");
                        rl_message.setVisibility(View.VISIBLE);
                        rl_message.postDelayed(new Runnable() {
                            @Override
                            public void run() {
                                Animation a = AnimationUtils.loadAnimation(Mobile_Recharge.this, R.anim.animation_down);
                                rl_message.setAnimation(a);
                                rl_message.clearAnimation();
                                rl_message.setVisibility(View.GONE);
                            }
                        }, 3000);
                    } else if (editext_number.getText().toString().length() < 10) {
                        textview_message.setText("Please enter a valid mobileno");
                        rl_message.setVisibility(View.VISIBLE);
                        rl_message.postDelayed(new Runnable() {
                            @Override
                            public void run() {
                                Animation a = AnimationUtils.loadAnimation(Mobile_Recharge.this, R.anim.animation_down);
                                rl_message.setAnimation(a);
                                rl_message.clearAnimation();
                                rl_message.setVisibility(View.GONE);
                            }
                        }, 3000);
                    }
                    else if (provider.equals("")) {
                        textview_message.setText("Please select operator");
                        rl_message.setVisibility(View.VISIBLE);
                        rl_message.postDelayed(new Runnable() {
                            @Override
                            public void run() {
                                Animation a = AnimationUtils.loadAnimation(Mobile_Recharge.this, R.anim.animation_down);
                                rl_message.setAnimation(a);
                                rl_message.clearAnimation();
                                rl_message.setVisibility(View.GONE);
                            }
                        }, 3000);
                    } else if (edittext_amount.getText().toString().equals("")) {
                        textview_message.setText("Please enter amount");
                        rl_message.setVisibility(View.VISIBLE);
                        rl_message.postDelayed(new Runnable() {
                            @Override
                            public void run() {
                                Animation a = AnimationUtils.loadAnimation(Mobile_Recharge.this, R.anim.animation_down);
                                rl_message.setAnimation(a);
                                rl_message.clearAnimation();
                                rl_message.setVisibility(View.GONE);
                            }
                        }, 3000);
                    }
                    else {
//
                         number = editext_number.getText().toString();
                         amount = edittext_amount.getText().toString();
                         mShowDialog("Please confirm operator, number and amount, wrong recharge will not reverse at any circumstances.");

                    }
                }
                else
                {
                    textview_message.setText("No Internet Connection");
                    rl_message.setVisibility(View.VISIBLE);
                    rl_message.postDelayed(new Runnable() {
                        @Override
                        public void run() {
                            Animation a = AnimationUtils.loadAnimation(Mobile_Recharge.this, R.anim.animation_down);
                            rl_message.setAnimation(a);
                            rl_message.clearAnimation();
                            rl_message.setVisibility(View.GONE);
                        }
                    }, 3000);
                }
            }
        });

    }


    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                finish();
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    private  void  mRechargeMobile(final  String appToken,final String user_id,final String provider,final String amount,final String number) {
        class getJSONData extends AsyncTask<String, String, String> {

            HttpURLConnection urlConnection;

            @Override
            protected void onPreExecute() {
                super.onPreExecute();
                dialog=new ProgressDialog(Mobile_Recharge.this);
                dialog.setMessage("Please wait...");
                dialog.show();
                dialog.setCancelable(false);
            }
            @Override
            protected String doInBackground(String... args) {

                StringBuilder result = new StringBuilder();

                try {
                    Log.e("url", SharePrfeManager.getInstance(Mobile_Recharge.this).mGetBaseUrl()+"api/android/recharge/pay?apptoken="+appToken+"&user_id="+user_id+"&provider_id="+provider+"&amount="+amount+"&number="+number);
                    URL url = new URL(SharePrfeManager.getInstance(Mobile_Recharge.this).mGetBaseUrl()+"api/android/recharge/pay?apptoken="+appToken+"&user_id="+user_id+"&provider_id="+provider+"&amount="+amount+"&number="+number);
                    urlConnection = (HttpURLConnection) url.openConnection();
                    InputStream in = new BufferedInputStream(urlConnection.getInputStream());

                    BufferedReader reader = new BufferedReader(new InputStreamReader(in));

                    String line;
                    while ((line = reader.readLine()) != null) {
                        result.append(line);
                    }

                } catch (Exception e) {
                    e.printStackTrace();
                } finally {
                    urlConnection.disconnect();
                }
                return result.toString();
            }

            @Override
            protected void onPostExecute(String result) {
                dialog.dismiss();

                Log.e("data",result);
                Showdata(result);
            }

        }

        getJSONData getJSONData=new getJSONData();
        getJSONData.execute();
    }

    public void GetData(String provider_id,String name)
    {
        operator_name=name;
        textview_operator.setText(operator_name);
        provider=provider_id;

    }

    protected void Showdata(final String json)
    {
        Log.e("response", json);
        String status="";
        String message="";
        String txnid="";
        String rrn="";
        try {
            JSONObject jsonObject = new JSONObject(json);
            status=jsonObject.getString("status");
            if (status.equalsIgnoreCase("ERR")){
                message=jsonObject.getString("message");
            }else {
                message = status;//jsonObject.getString("description");
            }
            if (jsonObject.has("txnid")){
                txnid=jsonObject.getString("txnid");
            }else{
                txnid= "Failed";
            }
            if (jsonObject.has("rrn")){
                rrn=jsonObject.getString("rrn");
            }else{
                rrn= "Failed";
            }
        }
        catch (JSONException e) {
            e.printStackTrace();
        }

        Intent intent=new Intent(Mobile_Recharge.this, TransactionReciept.class);
        intent.putExtra("status",status);
        intent.putExtra("message",message);
        intent.putExtra("transactionid", txnid);
        intent.putExtra("transactionrrn", rrn);
        intent.putExtra("transaction_type","Mobile Recharge");
        intent.putExtra("operator",textview_operator.getText().toString());
        intent.putExtra("number",number);
        intent.putExtra("price",amount);
        startActivity(intent);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        // TODO Auto-generated method stub
        super.onActivityResult(requestCode, resultCode, data);
        switch (requestCode) {
            case (10) :
                if (resultCode == Activity.RESULT_OK) {
                    Cursor cursor = null;
                    try {
                        String phoneNo = null ;
                        String name = null;
                        Uri uri = data.getData();
                        cursor = getContentResolver().query(uri, null, null, null, null);
                        cursor.moveToFirst();
                        int  phoneIndex =cursor.getColumnIndex(ContactsContract.CommonDataKinds.Phone.NUMBER);
                        phoneNo = cursor.getString(phoneIndex);
                        phoneNo=phoneNo.replaceAll(" ","");

                        if (phoneNo.startsWith("+91"))
                        {
                            editext_number.setText(phoneNo.substring(3,13));
                        }

                        else {
                            editext_number.setText(phoneNo);
                        }
                    } catch (Exception e) {
                        e.printStackTrace();
                    }
                }
                break;
            case (12):
                if (resultCode == Activity.RESULT_OK) {
                    Bundle res = data.getExtras();
                    String result = res.getString("qr_code");
                    editext_number.setText(result);

                }else if(resultCode==Activity.RESULT_CANCELED){
                    Toast.makeText(getApplicationContext(),"Unable to read QR Code.Please Try Again.",Toast.LENGTH_LONG).show();
                }
                break;
        }

    }

    protected void mShowDialog(final String message)
    {
        final AlertDialog.Builder builder1= new AlertDialog.Builder(Mobile_Recharge.this);
        builder1.setMessage("Recharge of "+operator_name + "\nAmount - \u20B9 "+amount + "\nNote - " + message) .setTitle("Recharge Confirmation");

        builder1.setCancelable(false)
                .setPositiveButton("Confirm", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int i) {
                        if (DetectConnection.checkInternetConnection(Mobile_Recharge.this)){
                            mRechargeMobile(SharePrfeManager.getInstance(Mobile_Recharge.this).mGetappToken(), SharePrfeManager.getInstance(Mobile_Recharge.this).mGetUserId(),provider,amount,number);
                            dialog.dismiss();
                        }
                        else
                        {
                            Toast.makeText(getApplicationContext(), "No Internet Connection", Toast.LENGTH_SHORT).show();
                        }
                    }
                })
                .setNegativeButton("Cancel", new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialog, int id) {
                        dialog.cancel();
                    }
                });
        AlertDialog alert = builder1.create();
        alert.show();
    }

    public void mCloseFragment()
    {
        getSupportFragmentManager().popBackStack();
    }


}

File: app/src/main/java/com/swipecare/payments/recharges/MobileRecharge.kt
package com.swipecare.payments.recharges

import android.app.Activity
import android.app.AlertDialog
import android.app.ProgressDialog
import android.content.Context
import android.content.Intent
import android.database.Cursor
import android.os.Bundle
import android.provider.ContactsContract
import android.view.MenuItem
import android.view.View
import android.view.animation.AnimationUtils
import android.view.inputmethod.InputMethodManager
import android.widget.RelativeLayout
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.swipecare.payments.*
import com.swipecare.payments.Operators.OperatorsFragment
import com.swipecare.payments.databinding.MobileRechargesBinding
import com.swipecare.payments.model.MobileRechargeResponse
import com.swipecare.payments.network.retrofit.SwipeApi
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext


class MobileRecharge : AppCompatActivity() {

    private lateinit var binding: MobileRechargesBinding
    private lateinit var mobileRechargeViewModel: MobileRechargeViewModel

    private lateinit var rlMessage: RelativeLayout

    private lateinit var username: String
    private lateinit var password: String
    private var balance: String = ""
    private var provider: String = ""

    private lateinit var progressDialog: ProgressDialog

    private val rechargePlanResult = 11
    private val rechargeOfferResult = 21

    var amount = ""
    var number: String? = null
    private lateinit var textviewMessage: TextView

    private lateinit var operatorName: String

    var icon = ""
    private lateinit var coroutineScope: CoroutineScope

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        // Inflate the layout for this fragment
        binding =
            MobileRechargesBinding.inflate(layoutInflater)
        setContentView(binding.root)

        progressDialog = ProgressDialog(this@MobileRecharge)
        progressDialog.setMessage("Processing your recharge.Please wait..")

        rlMessage = binding.rlMessage
        textviewMessage = binding.textviewMessage
        coroutineScope = CoroutineScope(Dispatchers.IO)

        supportActionBar!!.setDisplayHomeAsUpEnabled(true)
        binding.llDth.setOnClickListener {
            startActivity(Intent(this, DTH::class.java))
            this.overridePendingTransition(
                android.R.anim.slide_in_left,
                android.R.anim.slide_out_right
            )
            finish()
        }
        binding.llElectricity.setOnClickListener {
            startActivity(Intent(this, Electricity::class.java))
            this.overridePendingTransition(
                android.R.anim.slide_in_left,
                android.R.anim.slide_out_right
            )
            finish()
        }
        val sharedPreferences = getSharedPreferences("user", MODE_PRIVATE)
        username = sharedPreferences.getString("username", "") ?: ""
        password = sharedPreferences.getString("password", "") ?: ""
        balance = sharedPreferences.getString("balance", "") ?: ""
        binding.imageviewContact.setOnClickListener {
            val intent =
                Intent(Intent.ACTION_PICK, ContactsContract.CommonDataKinds.Phone.CONTENT_URI)
            startActivityForResult(intent, 10)
        }
        binding.llOperator.setOnClickListener {
            binding.edittextEnterNumber.clearFocus()
            hideKeyboard(this)

            val bottomSheetDialogFragment = OperatorsFragment()
            val bundle = Bundle()
            bundle.putString("type", "1")
            bundle.putString("activity", "mobile")
            bottomSheetDialogFragment.arguments = bundle
            supportFragmentManager.beginTransaction().add(R.id.content, bottomSheetDialogFragment)
                .addToBackStack(null).commit()
        }
        binding.seePlans.setOnClickListener {
            val mobileNumber = binding.edittextEnterNumber.text.toString()
            if (mobileNumber.length != 10) {
                Toast.makeText(this, "Mobile Number should be 10 digits", Toast.LENGTH_SHORT).show()
            } else {
                val intent = Intent(this, MobileRechargePlansActivity::class.java)
                intent.putExtra("mobileNumber", mobileNumber)
                startActivityForResult(intent, rechargePlanResult)
            }
        }
        binding.buttonPaynow.setOnClickListener {
            if (DetectConnection.checkInternetConnection(this)) {
                if (allInputFieldsAreValid()) {
                    mShowDialog()
                } else {
                    showErrorMessage()
                }
            } else {
                textviewMessage.text = getString(R.string.no_internet_connection)
                showErrorMessage()
            }
        }
        binding.seeOffers.setOnClickListener {
            if (DetectConnection.checkInternetConnection(this)) {
                if (mobileNumberInputIsValid()) {
                    val intent = Intent(this, RechargeOffersActivity::class.java)
                    intent.putExtra("mobileNumber", binding.edittextEnterNumber.text.toString())
                    startActivityForResult(intent, rechargeOfferResult)
                } else {
                    textviewMessage.text = getString(R.string.please_enter_a_valid_mobile_no)
                    showErrorMessage()
                }
            } else {
                textviewMessage.text = getString(R.string.no_internet_connection)
                showErrorMessage()
            }
        }
    }

    private fun mobileNumberInputIsValid(): Boolean {
        return binding.edittextEnterNumber.text.toString().length == 10
    }

    private fun allInputFieldsAreValid(): Boolean {
        return when {
            binding.edittextEnterNumber.text.toString() == "" -> {
                textviewMessage.text = getString(R.string.please_enter_mobile_number)
                showErrorMessage()
                false
            }
            binding.edittextEnterNumber.text.toString().length < 10 -> {
                textviewMessage.text = getString(R.string.please_enter_a_valid_mobile_no)
                showErrorMessage()
                false
            }
            /*binding.edittextEnterNumber.text.toString().length == 10 -> {
                // fetch the operator and circle

            }*/
            provider == "" -> {
                textviewMessage.text = getString(R.string.please_select_operator)
                showErrorMessage()
                false
            }
            binding.edittextAmount.text.toString() == "" -> {
                textviewMessage.text = getString(R.string.please_enter_amount)
                showErrorMessage()
                false
            }
            else -> {
                number = binding.edittextEnterNumber.text.toString()
                amount = binding.edittextAmount.text.toString()
                true
            }
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            android.R.id.home -> {
                finish()
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        when (requestCode) {
            10 -> if (resultCode == RESULT_OK) {
                var cursor: Cursor? = null
                try {
                    var phoneNo: String? = null
                    val name: String? = null
                    val uri = data?.data
                    cursor = contentResolver.query(uri!!, null, null, null, null)
                    cursor!!.moveToFirst()
                    val phoneIndex =
                        cursor.getColumnIndex(ContactsContract.CommonDataKinds.Phone.NUMBER)
                    phoneNo = cursor.getString(phoneIndex)
                    phoneNo = phoneNo.replace(" ".toRegex(), "")
                    if (phoneNo.startsWith("+91")) {
                        binding.edittextEnterNumber.setText(phoneNo.substring(3, 13))
                    } else {
                        binding.edittextEnterNumber.setText(phoneNo)
                    }
                } catch (e: Exception) {
                    e.printStackTrace()
                }
                cursor?.close()
            }
            12 -> if (resultCode == RESULT_OK) {
                val res = data?.extras
                val result = res!!.getString("qr_code")
                binding.edittextEnterNumber.setText(result)
            } else if (resultCode == RESULT_CANCELED) {
                Toast.makeText(
                    applicationContext,
                    "Unable to read QR Code.Please Try Again.",
                    Toast.LENGTH_LONG
                ).show()
            }
            rechargePlanResult -> {
                if (resultCode == RESULT_OK) {
                    val planAmount = data?.getStringExtra("planPrice")
                    if (planAmount?.isNotEmpty() == true) {
                        binding.edittextAmount.setText(planAmount)
                    }
                }
            }
            rechargeOfferResult -> {
                if (resultCode == RESULT_OK) {
                    val planAmount = data?.getStringExtra("offerPrice")
                    if (planAmount?.isNotEmpty() == true) binding.edittextAmount.setText(planAmount)
                }
            }
        }
    }

    private fun mShowDialog(message: String = "Please confirm mobile number, operator and amount. We cannot reverse Wrong Recharges under any circumstances.") {
        val builder1 = AlertDialog.Builder(this)
        builder1.setMessage("Recharge for: ${binding.edittextEnterNumber.text}\nOperator: $operatorName\nAmount: \u20B9 $amount\nNote: $message")
            .setTitle("Recharge Confirmation")
        builder1.setCancelable(true)
            .setPositiveButton("Confirm") { dialog, i ->
                if (DetectConnection.checkInternetConnection(this)) {
                    val sharedPreferences = SharePrfeManager.getInstance(this)
                    val appToken = sharedPreferences.mGetappToken()
                    val userId = sharedPreferences.mGetUserId()
                    var mobileRechargeResponse: MobileRechargeResponse

                    coroutineScope.launch {
                        try {
                            withContext(Dispatchers.Main) {
                                progressDialog.show()
                                progressDialog.setCancelable(false)
                            }
                            mobileRechargeResponse = SwipeApi.retrofitService.doMobileRecharge(
                                appToken,
                                userId,
                                provider,
                                amount,
                                number!!
                            )
                            withContext(Dispatchers.Main) {
                                progressDialog.dismiss()
                                val intent =
                                    Intent(this@MobileRecharge, TransactionReciept::class.java)
                                intent.putExtra("status", mobileRechargeResponse.status)
                                if (mobileRechargeResponse.status == "ERR") {
                                    intent.putExtra("message", mobileRechargeResponse.description ?: "some error occurred.please try again")
                                } else {
                                    intent.putExtra("message", mobileRechargeResponse.status)
                                }
                                intent.putExtra("transactionid", mobileRechargeResponse.txnid ?: "")
                                intent.putExtra("transactionrrn", mobileRechargeResponse.refno ?: "")
                                intent.putExtra("transaction_type", "Mobile Recharge")
                                intent.putExtra(
                                    "operator",
                                    binding.textviewOperator.text.toString()
                                )
                                intent.putExtra("number", number)
                                intent.putExtra("price", amount)
                                startActivity(intent)
                            }
                        } catch (e: Exception) {
                            withContext(Dispatchers.Main) {
                                progressDialog.dismiss()
                                Toast.makeText(
                                    this@MobileRecharge,
                                    "Error occurred: Something went wrong please try later or Same transaction allowed after 2 min",
                                    Toast.LENGTH_SHORT
                                ).show()



                            }
                        }
                    }
                } else {
                    Toast.makeText(applicationContext, "No Internet Connection", Toast.LENGTH_SHORT)
                        .show()
                }
            }
            .setNegativeButton(
                "Cancel"
            ) { dialog, id -> dialog.cancel() }
        val alert = builder1.create()
        alert.show()
    }

    private fun showErrorMessage() {
        rlMessage.visibility = View.VISIBLE
        rlMessage.postDelayed({
            val a = AnimationUtils.loadAnimation(
                this,
                R.anim.animation_down
            )
            rlMessage.animation = a
            rlMessage.clearAnimation()
            rlMessage.visibility = View.GONE
        }, 3000)
    }

    fun mCloseFragment() {
        supportFragmentManager.popBackStack()
    }

    fun getData(provider_id: String, name: String) {
        operatorName = name
        binding.textviewOperator.text = operatorName
        provider = provider_id
    }

    private fun hideKeyboard(activity: Activity) {
        try {
            val imm = getSystemService(Context.INPUT_METHOD_SERVICE) as? InputMethodManager
            imm?.hideSoftInputFromWindow(findViewById<View>(android.R.id.content).windowToken, 0)
        } catch (e: NullPointerException) {
        }
    }

}